Add centerVertical option to Modal

adhayward opened the pull request:
I was attempting to make my modals vertically centered as per the [bootstrap docs](https://getbootstrap.com/docs/4.4/components/modal/#vertically-centered) and discovered there was no way to add the appropriate class via this widget so I thought I'd add one.

This PR adds two new properties to the Modal widget:
1. centerVertical - a boolean that when true includes the modal-dialog-centered class. Defaults to false for BC

2. dialogOptions - an array for the options on the modal-dialog div. The classes that were previously added to this element are now added to dialogOptions via Html::addCssClass


| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | Y
| Fixed issues  |

simialbi commented:
I miss a test here

adhayward commented:
I've added tests now and also included an option for modal-dialog-scrollable too as that can easily work in the same way as modal-dialog-centered.

simialbi commented:
Nice. Ty. If you introduce `$dialogOptions` anyway, could you add the `role` parameter the same way it's done in `$options`?

simialbi commented on src/Modal.php:
```diff
@@ -125,9 +143,18 @@ public function init()
 
         $this->initOptions();
 
+        Html::addCssClass($this->dialogOptions, 'modal-dialog');
```
I'd move this to `initOptions` method.

simialbi commented on src/Modal.php:
```diff
@@ -125,9 +143,18 @@ public function init()
 
         $this->initOptions();
 
+        Html::addCssClass($this->dialogOptions, 'modal-dialog');
+        Html::addCssClass($this->dialogOptions,  $this->size);
+        if ($this->centerVertical) {
+            Html::addCssClass($this->dialogOptions,  'modal-dialog-centered');
+        }
+        if ($this->scrollable) {
+            Html::addCssClass($this->dialogOptions,  'modal-dialog-scrollable');
+        }
```
I'd move this to `initOptions` method.

adhayward commented:
@simialbi I've moved those lines to initOptions now and added the role="document" in.

The role addition does mean that the default output is different to before - does it count as breaking BC? As it's a correction to bring it inline with the bootstrap docs and shouldn't have any major effects I would assume it's ok but probably worth mentioning.

samdark commented:
Adding role is not a big deal in terms of breaking anything. It's fine to have it without major version bump. A mention in release notes is a good idea.

samdark commented on src/Modal.php:
```diff
@@ -115,7 +115,25 @@ class Modal extends Widget
+    /**
+     * @var boolean whether to center the modal vertically
+     *
+     * When true the modal-dialog-centered class will be added to the modal-dialog
```
Need `@since 2.0.9`

samdark commented on src/Modal.php:
```diff
@@ -115,7 +115,25 @@ class Modal extends Widget
+    /**
+     * @var boolean whether to make the modal body scrollable
+     *
+     * When true the modal-dialog-scrollable class will be added to the modal-dialog
```
Need `@since 2.0.9`

samdark commented on src/Modal.php:
```diff
@@ -115,7 +115,25 @@ class Modal extends Widget
+    public $scrollable = false;
 
+    /**
+     * @var array modal dialog options
```
Need `@since 2.0.9`

adhayward commented:
@samdark those should now be included

samdark commented:
Merged. Thank you!
